### PR TITLE
Add vecBCRdist to Amulety

### PR DESCRIPTION
**Pull request by wenggyaoo**

### CODE REVIEW

File: .github/workflows/vecbcrdist_reference.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        with:
+          repository: Kleinstein-Lab/convbench
+          ref: 65a425db711b9d6e94900d430c672e60098b8ab6
+          path: .reference/convbench
+      - name: Verify the ConvBench revision
+        run: test "$(git -C .reference/convbench rev-parse HEAD)" = "65a425db711b9d6e94900d430c672e60098b8ab6"
+      - name: Compare vecBCRdist with ConvBench BCRdist
+        env:
+          CONVBENCH_TCRDIST3_PATH: ${{ github.workspace }}/.reference/convbench/bin/tcrdist3.py
+        run: pytest -q tests/test_bcr_embeddings.py -k vec_bcrdist_matches_convbench_bcrdist
```
Comment (ggabernet) [inline .github/workflows/vecbcrdist_reference.yml:34]: Sorry not to be explicit enough here. We don't want to include the dependency to convbench inside the amulety package. I mean that we should test that they are equivalent manually, outside of amulety. Could the reference and import to convbench be removed from amulety?

File: .github/workflows/vecbcrdist_reference.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install pytest
+          python -m pip install -e ".[vecbcrdist-reference]"
+      - name: Checkout the fixed ConvBench reference
+        uses: actions/checkout@v4
+        with:
+          repository: Kleinstein-Lab/convbench
+          ref: 65a425db711b9d6e94900d430c672e60098b8ab6
+          path: .reference/convbench
```
Comment (ggabernet) [inline .github/workflows/vecbcrdist_reference.yml:28]:
```suggestion
```

File: .github/workflows/vecbcrdist_reference.yml (hunk 1 above)
Comment (ggabernet) [reply .github/workflows/vecbcrdist_reference.yml:34]: You might then not need this workflow at all, correct?

File: amulety/amulety.py [hunk 3]
```diff
@@ -406,7 +406,16 @@ def embed_airr(
 
     # ===== MODEL EXECUTION WITH CHAIN VALIDATION =====
     # BCR models
-    if model == "ablang":
+    if model == "vec-bcrdist":
```
Comment (ggabernet) [inline amulety/amulety.py:409]: let's add them alphabetically ordered so we can easily find all models :)

File: amulety/bcr_embeddings.py [hunk 4]
```diff
@@ -400,3 +404,199 @@ def balm_paired(
         residue_level=residue_level,
     )
     return embeddings
+
+
+# implement vexTCRdist from vecTCRdist:
+# Kept from clustcrdist.constants.hashing: the canonical amino-acid order and the BLOSUM62-derived tcrdist substitution-cost matrix.
+AALPHABET = "ARNDCQEGHILKMFPSTWYV"
+GAPCHAR = "-"
+BLOSUM_62 = np.array(
```
Comment (ggabernet) [inline amulety/bcr_embeddings.py:413]: Can we add this matrix in a separate file, called blosum_62.csv for example under a new directory amulety/matrices. And just read this one from file here?

File: amulety/bcr_embeddings.py [hunk 5]
```diff
@@ -400,3 +404,199 @@ def balm_paired(
         residue_level=residue_level,
     )
     return embeddings
+
+
+# implement vexTCRdist from vecTCRdist:
```
Comment (ggabernet) [inline amulety/bcr_embeddings.py:409]:
```suggestion
# implement vecBCRdist from vecTCRdist:
```

File: amulety/bcr_embeddings.py [hunk 6]
```diff
@@ -400,3 +404,199 @@ def balm_paired(
...
+    bcrs: pd.DataFrame,
+    cdr1_col: str = "cdr1_aa",
+    cdr2_col: str = "cdr2_aa",
+    cdr3_col: str = "cdr3_aa",
+):
+    """Replace V-gene lookup with observed BCR CDR1/2/3 strings."""
+    cdr_columns = (cdr1_col, cdr2_col, cdr3_col)
+    missing_columns = [column for column in cdr_columns if column not in bcrs.columns]
+    if missing_columns:
+        raise ValueError(f"Missing BCR CDR column(s): {missing_columns}")
```
Comment (ggabernet) [inline amulety/bcr_embeddings.py:451]:
```suggestion
        raise ValueError(f"Missing BCR CDR column(s): {missing_columns}. Make sure you run `amulety translate-igblast` with the option `--keep-regions` to obtain the CDR amino acid sequences.".)
```

File: amulety/bcr_embeddings.py [hunk 7]
```diff
@@ -400,3 +404,199 @@ def balm_paired(
...
+
+    allowed = set(AALPHABET)
```
Comment (ggabernet) [inline amulety/bcr_embeddings.py:453]: I have a question about the allowed aa alphabet and the gaps. For you and also @ZhaochenYe999.
1) should the allowed characters also include the gaps here, or is this handled below?
2) does vecTCRdist then expect the CDRs to be aligned with each other with IMGT gaps, if the gap characters are allowed? Do we perform this alignment in the BCR data?

I do see that the CDRs in the original `combo_xcr.tsv` are gapped. FOr example: TSG......FNG;VVL....DGL;SRSN.GY;CAVR..

File: amulety/bcr_embeddings.py [hunk 8]
```diff
@@ -400,3 +404,199 @@ def balm_paired(
...
+        self.dm[:20, :20] = self.distance_matrix
+        self.dm[:20, 20] = self.gap_penalty
+        self.dm[20, :20] = self.gap_penalty
+        if sqrt:
+            self.dm = np.sqrt(self.dm)
+        vectors = self._calc_mds_vecs()
+        return {amino_acid: vector for amino_acid, vector in zip(AALPHABET + GAPCHAR, vectors)}
+
+    @staticmethod
+    def _trim_and_gap(cdr: str, num_pos: int, n_trim: int = 0, c_trim: int = 0) -> str:
```
Comment (ggabernet) [inline amulety/bcr_embeddings.py:534]: @ZhaochenYe999 Are we sure that this is done properly here and that the gaps introduced are not specific for TCRs and can be used for the BCR CDR sequences?

File: amulety/bcr_embeddings.py [hunk 9]
```diff
@@ -400,3 +404,199 @@ def balm_paired(
...
+
+    def transform(self, x: pd.DataFrame) -> np.ndarray:
+        """Replace vecTCRdist's V-gene/junction DataFrame transform branch."""
+        check_is_fitted(self, ("aa_vectors_", "m"))
+        if not isinstance(x, pd.DataFrame):
+            raise TypeError("BCRDistEncoder expects an AIRR pandas DataFrame")
+        return self._gapped_encode_bcr_cdrs(x)
+
+
+def vec_bcrdist(cdr_sequences: pd.DataFrame, aa_dim: int = 8):
```
Comment (ggabernet) [inline amulety/bcr_embeddings.py:598]: Just for future work, could we easily also include `vec_tcrdist` inside amulety by including directly the class from the clustcrdist package?

File: amulety/utils.py [hunk 10]
```diff
@@ -528,6 +528,13 @@ def check_dependencies():
     except ImportError:
         missing_deps.append(("AbLang", "pip install ablang"))
 
+    try:
+        from sklearn.manifold import MDS  # noqa: F401
+
+        available_models.append("vecBCRdist")
+    except ImportError:
+        missing_deps.append(("vecBCRdist", "pip install scikit-learn"))
```
Comment (ggabernet) [inline amulety/utils.py:536]: If scikit-learn is already included in the dependencies list for amulety, it shouldn't be needed to install it via pip?

File: tests/test_bcr_embeddings.py [hunk 11]
```diff
@@ -234,3 +266,122 @@ def test_balm_paired_residue_level(self):
...
+        self.assertEqual(embeddings.dtype, torch.float32)
+        self.assertFalse(torch.equal(embeddings[0], embeddings[1]))
+
+    def test_vec_bcrdist_encoder_is_deterministic(self):
+        """The fixed-seed AA MDS encoding is reproducible."""
+        first = BCRDistEncoder().fit_transform(self.vec_bcrdist_df)
+        second = BCRDistEncoder().fit_transform(self.vec_bcrdist_df)
+        np.testing.assert_allclose(first, second)
+
+    @unittest.skipUnless(HAS_CONVBENCH_REFERENCE, "requires pinned ConvBench tcrdist3 reference dependencies")
```
Comment (ggabernet) [inline tests/test_bcr_embeddings.py:290]: Let's remove this unittest and just test this once manually with the dataset that Zhaochen shares with you

File: tests/test_bcr_embeddings.py [hunk 12]
```diff
@@ -234,3 +266,122 @@ def test_balm_paired_residue_level(self):
...
+        vectors = BCRDistEncoder().fit_transform(data)
+        upper_triangle = np.triu_indices(len(data), k=1)
+        reference_bcrdist = reference[upper_triangle]
+        squared_l2 = np.sum((vectors[:, None] - vectors[None, :]) ** 2, axis=2)[upper_triangle]
+
+        self.assertTrue(np.all(reference_bcrdist > 0))
+        self.assertGreater(np.corrcoef(reference_bcrdist, squared_l2)[0, 1], 0.98)
+        self.assertLess(np.mean(np.abs(reference_bcrdist - squared_l2) / reference_bcrdist), 0.1)
+
+    @unittest.skipUnless(HAS_PWSEQDIST, "requires optional dependency: pwseqdist")
```
Comment (ggabernet) [inline tests/test_bcr_embeddings.py:340]: This is to test that vecBCRdist generates the same results than TCRdist3? I like having this test in.

Comment (ggabernet) [reply tests/test_bcr_embeddings.py:340]: I'm just not sure if it should be exactly the case, because of the extra region that TCRdist considers between CDR2 and CDR3. Is this the case @ZhaochenYe999, could you comment on that and revise this test as well?

**Review by ggabernet — COMMENTED** [review]
Just a few comments, otherwise looks good to me. Let's merge it to the dev branch first

File: amulety/bcr_embeddings.py (hunk 7 above)
Comment (wenggyaoo) [reply amulety/bcr_embeddings.py:453]: Thank you for pointing out the question, and here is my understanding:

1. Gaps (-) are inserted later by _trim_and_gap for fixed-length encoding.

2. In vecTCRdist, combo_xcr.tsv already stores IMGT-gapped germline CDR1/CDR2(/CDR2.5). CDR3 is usually ungapped, and length gaps are handled by the encoder/metric. We replaced the germline V-gene lookup with observed CDRs (v2cdr=False), so we do not expect or allow pre-aligned IMGT gaps in the input.

File: tests/test_bcr_embeddings.py (hunk 12 above)
Comment (wenggyaoo) [reply tests/test_bcr_embeddings.py:340]: > This is to test that vecBCRdist generates the same results than TCRdist3? I like having this test in.

This remaining test is only a pwseqdist BCRdist sanity check (CDR1+CDR2+CDR3; pmhc/CDR2.5 weight 0), not a vecBCRdist and TCRdist3 equivalence test (agree to move the comparison of vecBCRdist and TCRdist3 outside Amulety).

File: amulety/bcr_embeddings.py (hunk 9 above)
Comment (wenggyaoo) [reply amulety/bcr_embeddings.py:595]: Good idea for a follow-up. For TCR we can try to wrap clustcrdist.encoding.TCRDistEncoder as an optional vec-tcrdist model in tcr_embeddings.py.

File: tests/test_bcr_embeddings.py (hunk 12 above)
Comment (ggabernet) [reply tests/test_bcr_embeddings.py:312]: ah ok perfect that was my question, so you set CDR2.5 a weight of 0, that makes sense!

File: amulety/bcr_embeddings.py (hunk 8 above)
Comment (zhaochenyeyaleedu) [reply amulety/bcr_embeddings.py]: This is also currently how TCRdist/BCRdist add gaps to cdr1/2 with` fixed_gappos=TRUE`.  I found it not making so much sense (even for TCR), bc in practice, you would almost always add gaps at the 6th posititon of a CDR. However, the TCRdist paper has no justification on the benefit of doing this. On the other hand, IMGT adds gaps always in the middle of a CDR (universal to BCR/TCR), which could potentially be an improvement for BCRdist/vecBCRdist.

Comment (zhaochenyeyaleedu) [reply amulety/bcr_embeddings.py]: Also, for cdr3 where `fixed_gappos=FALSE` in TCR/BCRdist, there will be a window for adding gaps [5, len(cdr)-5], and TCRdistr/BCRdist would choose the one producing a minimum distance towards a certain sequence. However, this is not feasible for vecTCR/BCRdist where we don't have such ONE SINGLE sequence to define the "minimum distance".  Therefore, both vecTCR/BCRdist actually use `fixed_gappos=TRUE`, different from original TCR/BCRdist.

File: amulety/bcr_embeddings.py (hunk 7 above)
Comment (zhaochenyeyaleedu) [reply amulety/bcr_embeddings.py]: Totally agree with what Wengyao explained. Just to add 1 thing, for BCRdist, anything that's not 20-AA characters are considered as a perfect match, and only gaps added by BCRdist itself will induce a penality. If IMGT-gaps are already added before running BCRdist, for example, "RA...DV" and "RA..KDV" will have 0 distance even if their actual lengths are different. That is because BCRdist will test if 2 strings have the same lengths and only when they are different, the gap adding and penalty functions will be called. Thus, BCRdist is designed for non-gapped sequences, though it wouldn't report an error when the sequences already have gaps or other characters.

File: tests/test_bcr_embeddings.py (hunk 12 above)
Comment (zhaochenyeyaleedu) [reply tests/test_bcr_embeddings.py]: The current test looks all good to me (BCRdist from convbench vs. vecBCRdist).
